### ISSUE TITLE
ysoserial payloads do not work on Linux targets

### ISSUE THREAD

**Issue by pedrib** [rapid7/metasploit-framework#13753]
## Steps to reproduce
1. Run exploit in #13604 on a Linux target
2. Watch it fail

## Expected behavior
We get a shell.

## Current behavior
Nothing happens.

### Metasploit version
dc576a51f860ca6526d69083d33bf643eb9a3dcd

### I installed Metasploit with:
- [ ] Kali package via apt
- [ ] Omnibus installer (nightly)
- [ ] Commercial/Community installer (from http://www.rapid7.com/products/metasploit/download.jsp)
- [X] Source install (please specify ruby version)
ruby 2.6.6p146 (2020-03-31 revision 67876) [x86_64-linux]

### OS
Debian 10

So it looks like the Linux ysoserial payloads are completely broken. I have tested the payloads in PR #13604 from ysoserial upstream and they work fine, but the Metasploit ones do not work at all.

Tagging @wvu-r7 since he asked me to do so, and @cdelafuente-r7 as he's handling the exploit PR. Sorry it took me so long to open this bug report!

**Comment by pedrib** [comment]
CC @rdomanski

**Comment by pedrib** [comment]
This is a key issue, it really breaks all ysoserial modules on Linux, so I'll leave it open!

**Comment by pedrib** [comment]
Sent a PR for a new module which is also affected by the same bug, PR #14654 

I have narrowed down the problem a bit. It seems that only cmd/unix/generic works. For example, take the module from #14654 and run it with cmd/unix/generic and cmd = 'cat /dev/zero'. You will see that 'cat /dev/zero' is executed on the target. 
However any other cmd/unix/* payloads fail to work. The mentioned module should work just fine with cmd/unix/reverse_python (as python is on the target), but it fails.

This cmd/unix/generic "workaround" might also work with [#13604](https://github.com/rapid7/metasploit-framework/pull/13604), but I haven't tested it.

**Comment by smcintyre-r7** [comment]
Alright, I took a look into this issue using the `multi/scada/inductive_ignition_rce` module as the test case. I chose this module because the vulnerable software is available online. Setup was easy on Ubuntu 20.04 x64.

I think this is an issue with escaping characters, specifically `"`. By updating the command to `"bash -c {echo,#{Rex::Text.encode_base64(cmd)}}|{base64,-d}|bash"` on Linux systems, I was able to successfully execute every payload I tested including `cmd/unix/reverse_netcat` and `cmd/unix/reverse_python_ssl`. I tried escaping the `"` character using `\` but was unsuccessful.

Next step would be to check if this issue affects the other gadget chains or not. Either way, it's not ideal that the caller would need to handle escaping the command, but knowing what all is affected would help guide implementing a fix.

**Comment by smcintyre-r7** [comment]
I also can't seem to generate payloads that contain `"` using the latest upstream ysoserial build.

<details>
<summary>Testing Output</summary>

```
17:11:40 Downloads java -jar ysoserial-master-138dc36bd2-1.jar CommonsBeanutils1 'python -c "FOO"' | xxd -g 1 
Error while generating or serializing payload
javassist.CannotCompileException: [source error] ) is missing
	at javassist.CtBehavior.insertAfter(CtBehavior.java:877)
	at javassist.CtBehavior.insertAfter(CtBehavior.java:792)
	at ysoserial.payloads.util.Gadgets.createTemplatesImpl(Gadgets.java:120)
	at ysoserial.payloads.util.Gadgets.createTemplatesImpl(Gadgets.java:102)
	at ysoserial.payloads.CommonsBeanutils1.getObject(CommonsBeanutils1.java:20)
	at ysoserial.GeneratePayload.main(GeneratePayload.java:34)
Caused by: compile error: ) is missing
	at javassist.compiler.Parser.parseArgumentList(Parser.java:1340)
	at javassist.compiler.Parser.parseMethodCall(Parser.java:1180)
	at javassist.compiler.Parser.parsePostfix(Parser.java:1036)
	at javassist.compiler.Parser.parseUnaryExpr(Parser.java:888)
	at javassist.compiler.Parser.parseBinaryExpr(Parser.java:775)
	at javassist.compiler.Parser.parseConditionalExpr(Parser.java:719)
	at javassist.compiler.Parser.parseExpression(Parser.java:699)
	at javassist.compiler.Parser.parseDeclarationOrExpression(Parser.java:591)
	at javassist.compiler.Parser.parseStatement(Parser.java:277)
	at javassist.compiler.Javac.compileStmnt(Javac.java:567)
	at javassist.CtBehavior.insertAfterAdvice(CtBehavior.java:892)
	at javassist.CtBehavior.insertAfter(CtBehavior.java:851)
	... 5 more
17:11:46 Downloads java -jar ysoserial-master-138dc36bd2-1.jar CommonsBeanutils1 'python -c FOO' | xxd -g 1 
17:11:46 Downloads java -jar ysoserial-master-138dc36bd2-1.jar CommonsBeanutils1 'python -c FOO' | xxd -g 1
00000000: ac ed 00 05 73 72 00 17 6a 61 76 61 2e 75 74 69  ....sr..java.uti
00000010: 6c 2e 50 72 69 6f 72 69 74 79 51 75 65 75 65 94  l.PriorityQueue.
00000020: da 30 b4 fb 3f 82 b1 03 00 02 49 00 04 73 69 7a  .0..?.....I..siz
00000030: 65 4c 00 0a 63 6f 6d 70 61 72 61 74 6f 72 74 00  eL..comparatort.
00000040: 16 4c 6a 61 76 61 2f 75 74 69 6c 2f 43 6f 6d 70  .Ljava/util/Comp
00000050: 61 72 61 74 6f 72 3b 78 70 00 00 00 02 73 72 00  arator;xp....sr.
00000060: 2b 6f 72 67 2e 61 70 61 63 68 65 2e 63 6f 6d 6d  +org.apache.comm
00000070: 6f 6e 73 2e 62 65 61 6e 75 74 69 6c 73 2e 42 65  ons.beanutils.Be
00000080: 61 6e 43 6f 6d 70 61 72 61 74 6f 72 e3 a1 88 ea  anComparator....
...
```
</details>

From the stack trace it looks like there's an error [here](https://github.com/frohoff/ysoserial/blob/138dc36bd23ac4ba7da5a8c89f1cd9b3f6a1e2e5/src/main/java/ysoserial/payloads/util/Gadgets.java#L117-L120) presumably with escaping characters. The problem with this is I can't generate a payload with a `"` to know how it *should* look to compare.

I [commented](https://github.com/frohoff/ysoserial/issues/147#issuecomment-768617646) on the upstream project's existing issue regarding bad characters in command.

**Comment by pedrib** [comment]
Hmm yes actually that is a common problem I used to have. I actually use a patch that it's in that comment for producing payloads with spaces on Linux. Sad to see this is still a problem, ysoserial appears to be very unmaintained.

**Comment by smcintyre-r7** [comment]
**TL;DR** the `modified_type` parameter needs to be set appropriately to avoid issues.

Alright, I don't think a blanket fix for this is going to be feasible for a few reasons:
1. The upstream YSoSerial project has an issue open for the characters that hasn't been addressed
2. It's not practical to patch in a serialized array of strings since the affected chains actually load a Java class blob, which is a more complicated format
3. There's an issue generating the new payload data set which I reported in #14731 that would need to be addressed if a YSoSerial-level patch addressed the issue

That brings us to the option of working around the issue. I wasn't able to think of a solution in which the command could be encoded for the necessary YSoSerial payloads in a platform-independent way. Now given that any solution I considered would require the platform information to be propagated to the serialization routine, we're looking at an API change. With this in mind, I think the simplest solution is to set the `modified_type` according to the target information. Moving forward this is going to be our best bet for working around these character restrictions in the various chains. I updated [wiki](https://github.com/rapid7/metasploit-framework/wiki/Generating-%60ysoserial%60-Java-serialized-objects) so that new contributions will have a solid reference that clearly mentions the `modified_type` parameter and its importance.

I submitted a fix implementing this change for the Inductive Ignition RCE module in #14732 which was the original use case that raised this issue. Once that's landed I think we'll be ready to close this and know that it's simply something we'll have to work around. With the wiki updates I made, we'll hopefully be in a more informed position.
